From 41fbddfc20569cff75c30b9c67db770b51451873 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sun, 17 Aug 2014 21:31:47 -0400
Subject: [pci] Fix unaligned access in AF transaction pending test

Message-id: <20140817213147.4495.83049.stgit@gir.stowe>
Patchwork-id: 87825
O-Subject: [RHEL7.1 PATCH 70/69] PCI: Fix unaligned access in AF transaction pending test
Bugzilla: 1110896
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110896
Upstream Status: d066c946a866268c14a120b33e7226e899981998

commit d066c946a866268c14a120b33e7226e899981998
Author: Alex Williamson <alex.williamson@redhat.com>
Date:   Tue Jun 17 15:40:13 2014 -0600

    PCI: Fix unaligned access in AF transaction pending test

    pci_wait_for_pending() uses word access, so we shouldn't be passing
    an offset that is only byte aligned.  Use the control register offset
    instead, shifting the mask to match.

    Fixes: d0b4cc4e3270 ("PCI: Wrong register used to check pending traffic")
    Fixes: 157e876ffe0b ("PCI: Add pci_wait_for_pending() (refactor pci_wait_for_pending_transaction())
    Reported-by: Ben Hutchings <ben@decadent.org.uk>
    Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
    Acked-by: Gavin Shan <gwshan@linux.vnet.ibm.com>
    CC: stable@vger.kernel.org  # v3.14+

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/pci.c b/drivers/pci/pci.c
index 104b134..50e58c7 100644
--- a/drivers/pci/pci.c
+++ b/drivers/pci/pci.c
@@ -3121,8 +3121,13 @@ static int pci_af_flr(struct pci_dev *dev, int probe)
  if (probe)
   return 0;
 
- /* Wait for Transaction Pending bit clean */
- if (pci_wait_for_pending(dev, pos + PCI_AF_STATUS, PCI_AF_STATUS_TP))
+ /*
+  * Wait for Transaction Pending bit to clear.  A word-aligned test
+  * is used, so we use the conrol offset rather than status and shift
+  * the test bit to match.
+  */
+ if (pci_wait_for_pending(dev, pos + PCI_AF_CTRL,
+     PCI_AF_STATUS_TP << 8))
   goto clear;
 
  dev_err(&dev->dev, "transaction is not cleared; "
-- 
1.7.1